ws/config.py

```diff
@@ -1,180 +1,232 @@
-        for item in items:
-            if item not in choices:
-                raise configargparse.ArgumentTypeError("unknown item: '{}' (available choices: {})".format(item, choices))
-        return items
-
-    return wrapped
-
-def getArgParser(subname=None, *args, **kwargs):
-    if subname is None:
-        _, _script = os.path.split(sys.argv[0])
-        subname, _ = os.path.splitext(_script)
-
-    # create config file parser
-    cfp = ConfigFileParser("site", subname)
-    kwargs["config_file_parser"] = cfp
-    kwargs["ignore_unknown_config_file_keys"] = True
-
-    # set brief usage parameter by default
-    kwargs.setdefault("usage", "%(prog)s [options]")
-    # TODO: the supplied description is merged with the one from configargparse, which looks very ugly, regardless of the used formatter
-    kwargs.setdefault("formatter_class", configargparse.RawDescriptionHelpFormatter)
-
-    ap = configargparse.ArgParser(*args, **kwargs)
-
-    # add config file and site arguments
-    ap.add_argument("-c", "--config", metavar="PATH", is_config_file_arg=True,
-            help="path to config file (default: %(default)s)")
-    ap.add_argument("--site",
-            help="sets the top-level section to be read from config files (default: %(default)s)")
+        raise argparse.ArgumentTypeError(
+            "cannot convert '{}'to boolean".format(string))
+
+
+def argtype_configfile(string):
+    """Convert `--config` argument to an absolute file path."""
+    dirname = os.path.dirname(string)
+    name, ext = os.path.splitext(os.path.basename(string))
+
+    # Configuration name was specified.
+    if not dirname and not ext:
+        config_dir = os.getenv('XDG_CONFIG_HOME',
+                               os.path.expanduser(CONFIG_DIR))
+        path = os.path.join(config_dir,
+                            "{}/{}.conf".format(PROJECT_NAME, string))
+    # Relational or absolute path was specified.
+    elif:
+        if ext != '.conf':
+            raise argparse.ArgumentTypeError(
+                "config filename must end with '.conf' suffix")
+        path = os.path.abspath(os.path.expanduser(string))
+
+    return path
+        
+def argtype_path(string):
+    """Convert a string to an absolute filesystem path."""
+    return os.path.abspath(os.path.expanduser(string))
+
+
+# ==============
+# Action classes
+# ==============
+
+class _ActionCheckPath(argparse.Action):
+
+    def __call__(self, parser, namespace, path, option_string=None):
+        if not os.path.lexists(path)
+            raise FileNotFoundError(
+                "no such file or directory: {}".format(path)
+        setattr(namespace, self.dest, path)
+    
+
+class _ActionCheckDirname(argparse.Action):
+
+    def __call__(self, parser, namespace, path, option_string=None):
+        dirname = os.path.dirname(path)
+        if not os.path.isdir(dirname)
+            raise FileNotFoundError(
+                "no such directory: {}".format(dirname))
```

Comment:
I don't think this should be checked with `action="check_dirname"`. For example, you cannot combine it with `nargs="+"` or `action="append"` or `action="extend"`.